### PR TITLE
Add docstring using Github Copilot

### PR DESCRIPTION
**Pull request by TomPham97**
Branch from PR #18 

I'm using Github Copilot to automatically fill in the docstrings for the methods and classes according to the [Google Python style guide](https://google.github.io/styleguide/pyguide.html). Ideally, this should synergize well with mkdocs as we've discussed.

### CODE REVIEW

File: src/diffusers_interpret/attribution.py [hunk 1]
```diff
@@ -9,15 +9,31 @@ def gradients_attribution(
     pred_logits: torch.Tensor,
     input_embeds: Tuple[torch.Tensor],
     attribution_algorithms: List[AttributionAlgorithm],
-    explanation_2d_bounding_box: Optional[Tuple[Tuple[int, int], Tuple[int, int]]] = None,
+    explanation_2d_bounding_box: Optional[Tuple[Tuple[int,
+                                                      int], Tuple[int, int]]] = None,
     retain_graph: bool = False
 ) -> List[torch.Tensor]:
+    """Compute gradients attribution.
```
Comment (JoaoLages) [inline src/diffusers_interpret/attribution.py:16]:
```suggestion
    """Computes primary attribution scores from output gradients with respect to inputs.

```

File: src/diffusers_interpret/attribution.py [hunk 2]
```diff
@@ -9,15 +9,31 @@ def gradients_attribution(
...
     retain_graph: bool = False
 ) -> List[torch.Tensor]:
+    """Compute gradients attribution.
+
+    Args:
+        pred_logits: The logits of the model's prediction.
+        input_embeds: The input embeddings.
+        attribution_algorithms: The attribution algorithms to use.
+        explanation_2d_bounding_box: The 2D bounding box to use for the explanation.
+        retain_graph: Whether to retain the computation graph.
```
Comment (JoaoLages) [inline src/diffusers_interpret/attribution.py:23]: We also need to put the typing in the docstring, right?
```suggestion
        pred_logits (torch.Tensor): The logits of the model's prediction.
        input_embeds (Tuple[torch.Tensor]): A tuple with all the input embeddings.
        attribution_algorithms (List[AttributionAlgorithm]): A list of the attribution algorithms to use. One algorithm per input in `input_embeds`.
        explanation_2d_bounding_box (Tuple[Tuple[int, int], Tuple[int, int]]): Tuple with the bounding box coordinates where the attributions were calculated for. The tuple is like (upper left corner, bottom right corner). Example: `((0, 0), (300, 300))`. Defaults to `None`, which means that the gradients are computed for the whole output image.
        retain_graph (bool): Whether to retain the computation graph. Defaults to `False`.

```

File: src/diffusers_interpret/attribution.py [hunk 3]
```diff
@@ -9,15 +9,31 @@ def gradients_attribution(
...
+
+    Returns:
+        The attribution maps.
```
Comment (JoaoLages) [inline src/diffusers_interpret/attribution.py:26]: I think we can refer to the return typing like this.
```suggestion
        List[torch.Tensor]: A list with the primary attributions for each `input_embeds`.

```

File: src/diffusers_interpret/explainer.py [hunk 4]
```diff
@@ -21,10 +21,21 @@
 
 class BasePipelineExplainer(ABC):
     """
-    Core base class to explain all DiffusionPipeline: text2img, img2img and inpaint pipelines
+    Core base class to explain all DiffusionPipeline: text2img, img2img and inpaint pipelines.
+
+    Attributes:
+        pipe (DiffusionPipeline): The DiffusionPipeline to explain.
+        verbose (bool): If True, print progress bars and other information.
+        gradient_checkpointing (bool): If True, use gradient checkpointing to save memory.
```
Comment (JoaoLages) [inline src/diffusers_interpret/explainer.py:29]:
```suggestion
        pipe (`DiffusionPipeline`): The DiffusionPipeline to explain.
        verbose (`bool`): If True, print progress bars and other information.
        gradient_checkpointing (`bool`): If True, use gradient checkpointing to save memory.

```

File: src/diffusers_interpret/explainer.py [hunk 5]
```diff
@@ -21,10 +21,21 @@
...
+        verbose (bool): If True, print progress bars and other information.
+        gradient_checkpointing (bool): If True, use gradient checkpointing to save memory.
     """
 
     def __init__(self, pipe: DiffusionPipeline, verbose: bool = True, gradient_checkpointing: bool = False) -> None:
+        '''
+        Args:
+            pipe (DiffusionPipeline): The DiffusionPipeline to explain.
+            verbose (bool): If True, print progress bars and other information.
+            gradient_checkpointing (bool): If True, use gradient checkpointing to save memory.
```
Comment (JoaoLages) [inline src/diffusers_interpret/explainer.py:37]:
```suggestion
            pipe (`DiffusionPipeline`): The DiffusionPipeline to explain.
            verbose (`bool`): If True, print progress bars and other information.
            gradient_checkpointing (`bool`): If True, use gradient checkpointing to save memory.

```

File: src/diffusers_interpret/explainer.py [hunk 6]
```diff
@@ -310,6 +343,12 @@ def _get_attributions(
 
     @property
     def special_tokens_attributes(self) -> Set[str]:
+        '''
+        Returns a set of special tokens that are not considered when calculating attributions.
+
+        Returns:
+            A set of special tokens that are not considered when calculating attributions.
```
Comment (JoaoLages) [inline src/diffusers_interpret/explainer.py:350]:
```suggestion
            Set[str]: A set of special tokens that are not considered when calculating attributions.

```

File: src/diffusers_interpret/explainer.py [hunk 7]
```diff
@@ -339,10 +378,25 @@ def gradient_checkpointing_disable(self) -> None:
...
+        '''
         raise NotImplementedError
 
     @abstractmethod
     def get_prompt_tokens_token_ids_and_embeds(self, prompt: Union[str, List[str]]) -> Tuple[List[List[str]], BatchEncoding, torch.Tensor]:
+        '''
+        Returns the tokens, token ids and embeddings of the prompt.
+
+        Returns:
+            A tuple containing the tokens, token ids and embeddings of the prompt.
```
Comment (JoaoLages) [inline src/diffusers_interpret/explainer.py:395]:
```suggestion
            Tuple[List[List[str]], BatchEncoding, torch.Tensor]: A tuple containing the tokens, token ids, and embeddings of the prompt.

```

File: src/diffusers_interpret/explainer.py [hunk 8]
```diff
@@ -425,14 +480,75 @@ def _mimic_pipeline_call(
...
+            The model checkpoint map location to be used.
+        model_checkpoint_force_map_location (`bool`):
+            Whether or not to force map location for the model checkpoint.
+        model_checkpoint_ignore_mismatched_sizes (`bool`):
+            Whether or not to ignore mismatched sizes for the model checkpoint.
+        model_checkpoint_strict (`bool`):
+            Whether or not to use strict for the model checkpoint.
+        model_checkpoint_return_unused_kwargs (`bool`):
+            Whether or not to return unused kwargs for the model checkpoint.
+        model_checkpoint_kwargs
```
Comment (JoaoLages) [inline src/diffusers_interpret/explainer.py:549]: These are actually the properties of `BasePipelineImg2ImgExplainer.pipe`, we can remove from here, right?

**Review by JoaoLages — COMMENTED** [review]
Great job! We can merge this one after solving the points I raised

File: src/diffusers_interpret/data.py [hunk 9]
```diff
@@ -184,5 +210,32 @@ class AttributionAlgorithm(ExplicitEnum):
 
 @dataclass
 class AttributionMethods:
-    tokens_attribution_method: Union[str, AttributionAlgorithm] = AttributionAlgorithm.GRAD_X_INPUT
-    pixels_attribution_method: Optional[Union[str, AttributionAlgorithm]] = AttributionAlgorithm.MAX_GRAD
\ No newline at end of file
+    '''
+    Args:
+        tokens_attribution_method (`AttributionAlgorithm`)
```
Comment (JoaoLages) [inline src/diffusers_interpret/data.py:215]: Do we need to put `AttributionAlgorithm`, seems like we shouldn't
```suggestion
        tokens_attribution_method (AttributionAlgorithm)

```
Same comment for the rest.
Example in the docs:
image

**Comment by TomPham97** [comment]
Just in case you ever need to add annotation block, check out [this docstring format](https://mkdocstrings.github.io/python/usage/#google-style-admonitions).

**Pull request closed without merging**